shard/sharder.go

```diff
@@ -0,0 +1,352 @@
+	targetPeer := s.assignedPeer
+	s.currentShard = make(map[string]*cid.Cid)
+	s.assignedPeer = peer.ID("")
+	s.byteThreshold = 0
+	s.byteCount = 0
+	var retStr string
+	b := api.BlockWithFormat{
+		Data:   shardNode.RawData(),
+		Format: "cbor",
+	}
+	logger.Debugf("Here is the serialized ipld: %x", b.Data)
+	err = s.rpcClient.Call(targetPeer, "Cluster", "IPFSBlockPut",
+		b, &retStr)
+	if err != nil {
+		return err
+	}
+
+	// Track shard node as a normal pin within cluster state
+	pinS := api.Pin{
+		Cid:                  shardNode.Cid(),
+		Allocations:          []peer.ID{targetPeer},
+		ReplicationFactorMin: 1,
+		ReplicationFactorMax: 1,
+	}.ToSerial()
+
+	return s.rpcClient.Call(targetPeer, "Cluster", "Pin", pinS, &struct{}{})
+}
+
+/* Open questions
+
+1. getAssignment()
+   How to decide how much space to give out?  There seems to be an infinity
+   of different ways to do this so ultimately I see this being configurable
+   perhaps keeping strategies in the allocator config.  Orthoganal to this
+   is deciding WHICH peer gets the next shard.  Currently this is the choice
+   the cluster pin allocator makes and of course there is a rabbit hole of
+   work to get different properties and guarantees from this decision too.
+   Perhaps these two decisions will be made as part of a more general
+   decision incorporating other things as well?
+
+   For now I am assuming the WHICH peer decision is made first (probably
+   the peer with most open disk space) and want a way to answer HOW MUCH
+   space do we take from this peer to store a shard?  A few basic ideas:
+
+   A. All nodes offer up some fraction of their space (say 2/3) during each
+      allocation.  The emptiest nodes are allocated first. The second
+      alloc gets 2/3 of remaining 1/3 = 2/6 of the total empty disk space
+      etc.
+   B. Same as A but saturates in a finite number of steps.  First allocate
+      2/3 of remaining. If peer is allocated again allocate remaining 1/3
+   C. All nodes offer up min space that any node can afford.  All nodes
+      are allocated before repeating allocations (this bleeds into the
+      WHICH node decision)
+
+  Note pseudo-code as it stands doesn't take into account the shard node
+  when keeping block put data within threshold.  Size can't be determined
+  beforehand because blocks may be of different sizes.
+
+  We will probably be reusing the pin allocator.  Where is the best place for
+  the HOW MUCH decision code, generally in an allocator or here in the
+  sharder responding to information received from informer calls?
+
+2. Shard node (cluster dag) format
+   In the current go-ipfs landscape downloading resolvers for non-standard
+   ipld formats (i.e. everything but unixfs, and I think dag-cbor) is
+   a difficult requirement to ask from nodes in the cluster.
+
+   For this reason I think we should use one of ipfs's standard ipld
+   formats.  Is dag-cbor ok to use? Probably preferrable over unixfs
+   protobuf if we can help it.
```

Comment:
yes, let's try get things working with dag-cbor, but make it in a way that we can switch the implementation easily.

Reply:
Correct me if I am wrong but as far as I can tell casting between ipld formats is not generally easy (or always possible).
However As it stands the dag-cbor impl is not too hard to swap out with dag-protobuf.  We could simply grab the links from the cbor shard node with `Links()` make a new protonode and call the Public facing merkledag/node:`AddRawLink`.  We could also construct `ipld.Links` directly from the cids in the sharder state and do this.

Reply:
I didn't mean that we can switch and have compatibility with what we had done before, I meant that we can switch for testing (from scratch) and then settle on what works best for the final solution.